api/include/opentelemetry/trace/trace_state.h

```diff
@@ -97,37 +107,163 @@ class TraceState
+    }
+    return ts;
+  }
+
+  /**
+   * Creates a w3c tracestate header from TraceState object
+   */
+  std::string ToHeader()
   {
-    for (const auto &entry : Entries())
+    std::string header_s;
+    for (size_t count = 0; count < num_entries_; count++)
     {
+      if (count != 0)
+      {
+        header_s.append(",");
+      }
+      
+      auto entry = (entries_.get())[count];
+      header_s.append(entry.GetKey());
+      header_s.append(kKeyValueSeparator);
+      header_s.append(entry.GetValue());
+    }
+    return header_s;
+  }
+
+  /**
+   *  Returns `value` associated with `key` passed as argument
+   *  Returns empty string if key is invalid  or not found
+   */
+  std::string Get(nostd::string_view key) const noexcept
+  {
+    if (!IsValidKey(key))
+    {
+      return std::string();
+    }
+    for (size_t i = 0; i < num_entries_; i++)
+    {
+      auto entry = (entries_.get())[i];
       if (key == entry.GetKey())
       {
-        value = entry.GetValue();
-        return true;
+        return std::string(entry.GetValue());
       }
     }
-    return false;
+    return std::string();
   }
 
-  // Creates an Entry for the key-value pair and adds it to entries. Returns true if pair was added
-  // successfully, false otherwise. If value is null or entries_ is full, this function is a no-op.
-  bool Set(nostd::string_view key, nostd::string_view value) noexcept
+  /**
+   * Returns `new` TransState object with following mutations applied to the existing instance:
+   *  Update Key value: The updated value must be moved to beginning of List
+   *  Add : The new key-value pair SHOULD be added to beginning of List
+   *
+   * If the provided key-value pair is invalid, or results in transtate that violates the
+   * tracecontext specification, empty tracestate will be returned.
+   */
+  TraceState Set(const nostd::string_view &key, const nostd::string_view &value)
   {
-    if (value.empty() || num_entries_ >= kMaxKeyValuePairs)
+    TraceState ts;
+    if ((!IsValidKey(key) || !IsValidValue(value)) || num_entries_ == kMaxKeyValuePairs)
     {
-      return false;
+      // max size reached or invalid key/value. Returning empty tracestate
+      return ts;  // empty instance
     }
 
-    Entry entry(key, value);
-    (entries_.get())[num_entries_] = entry;
-    num_entries_++;
-    return true;
+    // add new key-value pair at beginning
+    Entry e(key, value);
+    (ts.entries_.get())[ts.num_entries_++] = e;
+    for (size_t i = 0; i < num_entries_; i++)
+    {
+      auto entry = (entries_.get())[i];
+      auto key   = entry.GetKey();
+      auto value = entry.GetValue();
+      Entry e(key, value);
+      (ts.entries_.get())[ts.num_entries_++] = e;
+    }
+    return ts;
+  }
+
+  /**
+   * Returns `new` TransState object after removing the attribute with given key ( if present )
+   * @returns empty TransState object if key is invalid
+   * @returns copy of original TransState object if key is not present (??)
```

Comment:
```suggestion
   * @returns copy of original TraceState object if key is not present (??)
```

Seems for now an empty instance is returned if the key is not present, should we return original `TraceState` like mentioned in the comment?

Reply:
Thanks for suggestion. Have changed it now as part of other changes.
Also, 
`empty instance is returned if the key is not present,`
 - As of now, as per code it return empty instance if key is not valid, and copy of original instance if key is not present.